python/tchannel/statsd.py

```diff
@@ -0,0 +1,143 @@
+            extract_meta_info_from_request(request))
+        key = '.'.join([statsd_name,
+                        clean(service, 'service'),
+                        clean(target_service, 'target-service'),
+                        clean(target_endpoint, 'target-endpoint'),
+                        clean(ErrorMessage.ERROR_CODES.
+                              get(error.code, None), 'type')
+                        ])
+
+        self._statsd.count(key, 1)
+
+    def after_receive_system_error_per_attempt(self, request, error):
+        statsd_name = "tchannel.outbound.calls.per-attempt.system-errors"
+        (service, target_service, target_endpoint) = (
+            extract_meta_info_from_request(request))
+        key = '.'.join([statsd_name,
+                        clean(service, 'service'),
+                        clean(target_service, 'target-service'),
+                        clean(target_endpoint, 'target-endpoint'),
+                        clean(ErrorMessage.ERROR_CODES.
+                              get(error.code, None), 'type')
+                        ])
+
+        self._statsd.count(key, 1)
+
+    def on_operational_error_per_attempt(self, request, error):
+        statsd_name = "tchannel.outbound.calls.per-attempt.operational-errors"
+        (service, target_service, target_endpoint) = (
+            extract_meta_info_from_request(request))
+        key = '.'.join([statsd_name,
+                        clean(service, 'service'),
+                        clean(target_service, 'target-service'),
+                        clean(target_endpoint, 'target-endpoint'),
+                        clean(ErrorMessage.ERROR_CODES.
+                              get(error.code, None), 'type')
+                        ])
+
+        self._statsd.count(key, 1)
+
+    def on_operational_error(self, request, error):
+        statsd_name = "tchannel.outbound.calls.operational-errors"
+        (service, target_service, target_endpoint) = (
+            extract_meta_info_from_request(request))
+        key = '.'.join([statsd_name,
+                        clean(service, 'service'),
+                        clean(target_service, 'target-service'),
+                        clean(target_endpoint, 'target-endpoint'),
+                        clean(ErrorMessage.ERROR_CODES.
+                              get(error.code, None), 'type')
+                        ])
+
+        self._statsd.count(key, 1)
+
+
+def extract_meta_info_from_request(request):
+    service = request.headers.get('cn', None)
+    target_service = request.service
+    target_endpoint = request.endpoint
+
+    return (service, target_service, target_endpoint)
+
+
+def clean(key, field):
+    if not key:
+        return 'no-' + field
+    else:
+
+        return key.replace(
+            ".", "-").replace(
+            "\/", "-").replace(
+            "{", "-").replace(
+            "}", "-").replace(
+            ":", "-").replace(
+            " ", "-")
```

Comment:
Use a regex-based substitution. 

Precompile it and put it in a constant

    re.compile(r'[{}/\\:\s.]')

And use [`.sub`](https://docs.python.org/2/library/re.html#re.sub) to replace all occurrences with `-`.

An advantage of this is that you can add a `+` to the character class and make it so that multiple adjacent occurrences of the special symbols are replaced with a single dash instead of two.

    >>> r = re.compile(r'[{}/\\:\s.]+')
    >>> r.sub('-', 'foo{}bar')
    'foo-bar'